src/thor/timedistancematrix.cc

```diff
@@ -67,12 +67,45 @@ void TimeDistanceMatrix::clear() {
   edgestatus_.clear();
 }
 
+// Initializes the time of the expansion if there is one
+std::vector<TimeInfo> TimeDistanceMatrix::SetTime(
+    google::protobuf::RepeatedPtrField<valhalla::Location>& origins,
+    const google::protobuf::RepeatedPtrField<valhalla::Location>& destinations,
+    GraphReader& reader,
+    float max_timedep_dist) {
+  // loop over all locations setting the date time with timezone
+  std::vector<TimeInfo> infos;
+  for (auto& origin : origins) {
+    // invalid time infos if the service doesn't allow it anyways
+    if (!max_timedep_dist || origin.date_time().empty()) {
+      infos.emplace_back(TimeInfo::invalid());
+      continue;
+    }
+    // only check distance to all destinations if a valid time object is set and the service allows it
+    PointLL origin_ll{origin.ll().lng(), origin.ll().lat()};
+    bool time_allowed = true;
+    for (const auto& destination : destinations) {
+      time_allowed = max_timedep_dist >
+                     origin_ll.Distance(PointLL(destination.ll().lng(), destination.ll().lat()));
+      if (!time_allowed) {
+        break;
+      }
+    }
+    infos.emplace_back(time_allowed ? TimeInfo::make(origin, reader, &tz_cache_)
+                                    : TimeInfo::invalid());
+  }
```

Comment:
here, we check if we're allowed to track time, based on the max distance between any pair. this sucks a bit.. in worst case this calculates origin_size x target_size distances.. 

the vector of `time_infos` has one element per origin (either source or target), if `max_timedep_dist` for that origin is fine.

instead of this "monster" we could also have a boolean in the config, "allow_time_dependent" matrix or so, and we wouldn't have to mess distances and just respect the normal `max_matrix_distance`

Reply:
we already check all pairs distances in loki and we always do the service limits checks there anyway. can we move this to loki? it would be nice if all this did was initialize the timeinfo stuff. in loki we can unset the time stuff if the service constraint fails and then here we'll see there is no time to set anyway.

Reply:
> we already check all pairs distances in loki

oops, I didn't even look :sweat_smile: I just saw how the matrix uses `max_matrix_distance` where it aborts the expansion if it's exceeded and thought that's the only way we use it, which did seem weird haha good point!